openshift-knative/serverless-operator
# Delete KnativeKafka resources when CR is deleted

Bases https://github.com/openshift-knative/serverless-operator/pull/487

Similar to other PRs in the umbrella ticket, https://github.com/openshift-knative/serverless-operator/issues/486, I didn't want to wait until the CI is green and that's merged. This is a WIP PR to see in advance if things fail and if possibly get some reviews.

This PR deletes the resources when the KnativeKafka CR is deleted.
Specifically,
- KafkaChannel resources --> garbage collection also works for these since they're in the same namespace and we set ownerRefs for them
- KafkaSource resources that are put in another namespace, `knative-sources` -->  no GC, manual deletion for these
- Deletion of cluster-scoped resources

Deletion of the resources when a component gets disabled is to be handled in another PR (e.g. `spec.channel.enabled=true` is changed to `spec.channel.enabled=false`)

As written in the umbrella ticket,  https://github.com/openshift-knative/serverless-operator/issues/486, there's still a lot to do.

Verification:

1. Enable the controller by uncommenting the stuff in `knative-operator/pkg/controller/add_knativekafka.go`

2. Install operator and eventing by `make install`

3. Need to manually create the CRD since it is not in the CSV yet
```
kubectl apply -f knative-operator/deploy/crds/operator_v1alpha1_knativekafka_crd.yaml
```

4. Install:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: true
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

5. Delete the CR and check if created resources are deleted:
```
kubectl delete knativekafkas example -n knative-eventing
```

6. Following commands should say `Error from server (NotFound): error when deleting "knative-operator/deploy/resources/knativekafka/kafka....yaml": foo "bar" not found` for each of the resource defined in them
```
k delete -f knative-operator/deploy/resources/knativekafka/kafkasource-latest.yaml
k delete -f knative-operator/deploy/resources/knativekafka/kafkachannel-latest.yaml
```
